### ISSUE TITLE
github actions cache usage allows invalid cache keys to be used

### ISSUE THREAD

**Issue by piersy** [clearmatics/autonity#504]
# Problem

It looks like our github actions caching is configured to allow use to restore code from a cache that does not relate to the branch being processed. Obviously we wan't to only be testing the code of the  branch which was pushed, not other branches.

From the [docs](https://help.github.com/en/actions/configuring-and-managing-workflows/caching-dependencies-to-speed-up-workflows#using-the-cache-action)
>You can optionally provide a list of restore-keys to use when the key doesn't match an existing cache. A list of restore-keys is useful when you are restoring a cache from another branch because restore-keys can partially match cache keys. For more information about matching restore-keys, see "Matching a cache key."

In all our uses of caching we provide the following restore key:
https://github.com/clearmatics/autonity/blob/62e991c62d4e5c5747159478b6734141d3d68ad7/.github/workflows/go.yml#L15

Which looks like it allows restoring a cache from any branch, if you consider that our full cache key definition is:
https://github.com/clearmatics/autonity/blob/62e991c62d4e5c5747159478b6734141d3d68ad7/.github/workflows/go.yml#L14 

# Implementation

Remove all use of restore keys.

# Acceptance Criteria

Github actions for a push  should only be run with the code of the commit that was pushed.

**Comment by JekaMas** [comment]
It will break linters.

**Comment by JekaMas** [comment]
Please check any CI results, eg https://github.com/clearmatics/autonity/pull/472/checks?check_run_id=564564727

You can check that it restores the code that is needed.

**Comment by piersy** [comment]
I don't understand your point @JekaMas?

The run you referenced seems to exhibit the behaviour that I'm concerned about. That being different code caches being restored.

The bootstrap job restores this cache key:
```
Linux-aut-cache-autonity-code-3d867b8e765c09ed01725668f130eb18e039946a
```
And the contract tests job restores this cache key:
```
Linux-aut-cache-autonity-code-907b9c5a16f87fcb6a6816109bbafd25260fc908
```

There is no guarantee that a cache is going to be available, which is why the restore keys are provided as a fallback.

Looking a bit deeper though I see that restore keys are only used when a cache miss is encountered, and since we always check out the code in the case of a cache miss, we will end up overwriting code retrieved from an invalid cache anyway. This will work out most of the time, but could cause a failure in the case that we restore an old cache that contains code that has since been deleted and now may not compile.

**Comment by piersy** [comment]
this has been fixed by #550
